### PR TITLE
[REVIEW] Select and Concat for cudf exp table

### PR DESCRIPTION
**Pull request by kaatish**
Addresses #3043 
Added select and concat for cudf::experimental::table

### CODE REVIEW

**Comment by harrism** [comment]
The term `concat` is already used for concatenating (for example) multiple same-type columns together into a single column. E.g. a vertical concat. This PR introduces a horizontal concat, which is potentially confusing. We may want to support table::concat() in the row axis. Therefore, I think we need to clarify, or add an axis parameter (like Pandas). Please discuss.

**Comment by jrhemstad** [comment]
How about:
```
std::unique_ptr<column> column_concat(std::vector<column_view> const& columns_to_concat);

table_view table_concat(std::vector<table_view> const& tables_to_concat);
```

> We may want to support table::concat() in the row axis. Therefore, I think we need to clarify, or add an axis parameter (like Pandas). Please discuss.

For a row-wise concat, I think we'd want to name something like `elementwise_concat` or `rowwise_concat`.

**Comment by harrism** [comment]
Why not just have an axis parameter?

**Comment by jrhemstad** [comment]
> Why not just have an axis parameter?

A horizontal concat vs. a vertical concat are fundamentally different operations that require incompatible return types. 

A "horizontal" concat is a lightweight operation that returns a `table_view`. It simply creates a table of `column_view`s into the passed in tables. No new allocations nor copies are required.

A "vertical" concat is a heavyweight operation that returns a `column` (or `table` for a vertical table concat). New `column`s must be allocated and filled containing copies of the input columns.

**Comment by harrism** [comment]
OK in that case I agree with something like what Jake suggested.

File: cpp/include/cudf/table/table.hpp [hunk 1]
```diff
@@ -100,9 +100,29 @@ class table {
    *---------------------------------------------------------------------------**/
   std::vector<std::unique_ptr<column>> release();
 
+  /**---------------------------------------------------------------------------*
+   * @brief Returns a table_view with set of specified columns.
+   *---------------------------------------------------------------------------**/
+  table_view select(std::vector<gdf_size_type> const& column_indices) const;
```
Comment (trevorsm7) [inline cpp/include/cudf/table/table.hpp]: Use cudf::size_type instead

File: cpp/src/table/table.cpp [hunk 2]
```diff
@@ -79,5 +79,25 @@ std::vector<std::unique_ptr<column>> table::release() {
   return std::move(_columns);
 }
 
+// Returns a table_view with set of specified columns
+table_view table::select(std::vector<gdf_size_type> const& column_indices) const {
```
Comment (trevorsm7) [inline cpp/src/table/table.cpp]: cudf::size_type here as well

File: cpp/src/table/table.cpp [hunk 3]
```diff
@@ -79,5 +79,25 @@ std::vector<std::unique_ptr<column>> table::release() {
...
+    CUDF_EXPECTS(column_indices.size() <= _columns.size(), "Requested too many columns.");
+
+    std::vector<column_view> columns;
+    for (auto index : column_indices) {
+      columns.push_back(_columns.at(index)->view());
```
Comment (jrhemstad) [inline cpp/src/table/table.cpp]: Update doc to indicate that this throws if `column_indices` has any index outside of `[0, num_columns())`.

Comment (kaatish) [reply cpp/src/table/table.cpp]: Updated [here](https://github.com/rapidsai/cudf/pull/3065/files#diff-6e2840973c2915fb2c562344d98e0886R120).

File: cpp/tests/table/table_tests.cu [hunk 4]
```diff
@@ -14,18 +14,138 @@
...
+using TView       = cudf::table_view;
+using Table       = cudf::experimental::table;
+
+struct TableViewTest : public cudf::test::BaseFixture {};
+
+template <typename T>
+struct TableTest : public cudf::test::BaseFixture {
+  std::default_random_engine generator;
+  std::uniform_int_distribution<int> distribution{1000, 10000};
+  int random_size() { return distribution(generator); }
```
Comment (jrhemstad) [inline cpp/tests/table/table_tests.cu]: If you inherit from the `UniformRandomGenerator` base class in `base_fixture.hpp` it handles the boilerplate for generating random numbers for you.

File: cpp/tests/table/table_tests.cu [hunk 5]
```diff
@@ -14,18 +14,138 @@
...
+};
+
+using TestingTypes = ::testing::Types<int8_t, int16_t, int32_t, int64_t, float,
```
Comment (jrhemstad) [inline cpp/tests/table/table_tests.cu]:
```suggestion
using TestingTypes = cudf::test::NumericTypes;
```

Comment (jrhemstad) [reply cpp/tests/table/table_tests.cu]: It doesn't look like `TypeParam` is ever being used in any of the `TYPED_TEST`s.

File: cpp/tests/table/table_tests.cu [hunk 6]
```diff
@@ -14,18 +14,138 @@
  * limitations under the License.
  */
 
+//#include <cudf/column/column.hpp>
```
Comment (jrhemstad) [inline cpp/tests/table/table_tests.cu]: If possible, this file should be made a .cpp. I don't see any device code in here, so I don't think it should be a problem.

File: cpp/tests/table/table_tests.cu (hunk 5 above)
Comment (kaatish) [reply cpp/tests/table/table_tests.cu]: I have removed all typed tests

File: cpp/tests/table/table_tests.cu (hunk 4 above)
Comment (kaatish) [reply cpp/tests/table/table_tests.cu]: I realized I do not need the random number generator in the test. Removed.

**Pull request merged**